Add java sol to closing the farm (silver) (#1387)
cpinitiative/usaco-guide: solutions/USACO open16-closing-the-farm.mdx

If all the unclosed nodes have been visited, then we print "YES"; otherwise, we
print "NO".

## Implementation

> **Time Complexity:** $\mathcal{O}(N^2)$

Reviewer: ```suggestion
**Time Complexity:** $\mathcal{O}(N^2 + NM)$
```

</CPPSection>

<JavaSection>

> ```java

Reviewer: use tabs instead of spaces to indent

Author: Fixed this just now, can you check if it's right?

Reviewer: Yep, it looks good now.

import java.util.*;
import java.io.*;

public class closing {
    //Store global variables used by dfs
    static ArrayList<ArrayList<Integer>> adj;
    static boolean[] visited, closed;
    static int nodes;

    public static void main(String[] args) throws IOException {
    	Scanner sc = new Scanner(new File("closing.in"));
    	PrintWriter out = new PrintWriter("closing.out");
    	
        int n = sc.nextInt();
        int m = sc.nextInt();

        visited = new boolean[n];
        closed = new boolean[n];
        adj = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<Integer>());

        for (int i = 0; i < m; i++) {

>             int a = sc.nextInt()-1;
>             int b = sc.nextInt()-1;

Reviewer: ```suggestion
            int a = sc.nextInt() - 1;
            int b = sc.nextInt() - 1;
```

            adj.get(a).add(b);
            adj.get(b).add(a);
        }

        int[] order = new int[n];
        for (int i = 0; i < n; i++) {

>             order[i] = sc.nextInt()-1;

Reviewer: ```suggestion
            order[i] = sc.nextInt() - 1;
```

        }

        for (int i = 0; i < n; i++) {
            Arrays.fill(visited, false);
            nodes = 0;
            

>             dfs(order[n-1]); // DFS from the barn that will close last
>             if (nodes == n-i) { // Checks if all unclosed barns have been visited

Reviewer: ```suggestion
            dfs(order[n - 1]); // DFS from the barn that will close last
            // Checks if all unclosed barns have been visited
            if (nodes == n - i) {
```

                out.println("YES");
            } else {
                out.println("NO");
            }
            closed[order[i]] = true;
        }
        out.close();
    }

    static void dfs(int node) {
        if (visited[node] || closed[node]) return;

>         // Increase the number of nodes if and only if they haven't been closed or visited previously

Reviewer: this line is too long

    	PrintWriter out = new PrintWriter("closing.out");
    	
        int n = sc.nextInt();
        int m = sc.nextInt();

        visited = new boolean[n];
        closed = new boolean[n];
        adj = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<Integer>());

        for (int i = 0; i < m; i++) {
            int a = sc.nextInt() - 1;
            int b = sc.nextInt() - 1;
            adj.get(a).add(b);
            adj.get(b).add(a);
        }

        int[] order = new int[n];
        for (int i = 0; i < n; i++) {
            order[i] = sc.nextInt() - 1;
        }

        for (int i = 0; i < n; i++) {
            Arrays.fill(visited, false);
            nodes = 0;
            
            dfs(order[n - 1]); // DFS from the barn that will close last
            // Checks if all unclosed barns have been visited
            if (nodes == n - i) {
                out.println("YES");
            } else {
                out.println("NO");
            }
            closed[order[i]] = true;
        }
        out.close();
    }

    static void dfs(int node) {
        if (visited[node] || closed[node]) return;

>         // Increase the number of nodes iff they haven't already been closed or visited

Reviewer: ```suggestion
        // Increase the number of nodes 
        // iff this barn hasn't already been closed or visited
```
still too long...maybe split it into two lines like this

</CPPSection>

<JavaSection>

```java
import java.util.*;
import java.io.*;

public class closing {

> 	//Store global variables used by dfs

Reviewer: ```suggestion
	// Store global variables used by dfs
```
to be consistent with the other comments
